hardening: fix #18995

cstrahan opened the pull request:
###### Motivation for this change

This should (hopefully) fix #18995.

This moves `hardeningLDFlags` to `extraBefore`, so e.g. user supplied `-O` flags should take precedence over the hardening flags. (EDIT: I'm actually not sure how I feel about this; [see commentary below](https://github.com/NixOS/nixpkgs/pull/28029#issuecomment-320921508))

Also, this changes the wrappers such that they now expect a whitelist of hardening flags, instead of (predominately) expecting a blacklist. Instead of taking `$hardeningEnable` and `$hardeningDisable` from the environment, they now use `$NIX_HARDENING_ENABLE` instead. If the user has set `hardeningEnable` and or `hardeningDisable` in a call to `mkDerivation`, we pass along the appropriate `NIX_HARDENING_ENABLE`. `stdenv/setup.sh` now sets `$NIX_HARDENING_ENABLE` to the default enabled hardening flags **_if_** the variable is unset.

This latter change is desirable for a couple reasons:

* Most wrapper configuration in Nix works via environment variables starting with `NIX_`; `hardening{Enable,Disable}` are the exceptions.
* When `gcc` et al are installed system-wide, most users don't realize that they have to explicitely opt out of hardening, which results in plenty of confusion and time lost trying to figure out why/when things are mysteriously broken.

This should provide the same behavior as before, aside from fixing the unintuitive bits mentioned above.

Note that I left a bit of `add-hardening.sh` oddly indented; I wanted to keep the diff readable during review. I'll fix the indentation before we merge.

If it makes sense, I can move the derivation of `$NIX_HARDENING_ENABLE` from Nix to `stdenv/setup.sh`, so that setting `hardening{Enable,Disable}` from Nix in a regular old `derivation` will work just as well as from `mkDerivation` (right now, that'll only happen for the latter). Please let me know how feel about this, either way.

Feel free to ask me any questions, or point out any concerns.

~~I'm still testing this at the moment.~~ This now works as described. See comments below for any open questions.

###### Things done

Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers.

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cstrahan committed:
hardening: fix #18995

Ericson2314 commented:
How is all handled now? (Granted looking on phone.)

CC @orivej

Ericson2314 commented:
Nevermind did not scroll down enough. I like this: Nix > bash.

orivej commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -3,6 +3,9 @@ set -o pipefail
 : ${outputs:=out}
 
+# If unset, assume the default hardening flags.
+: ${NIX_HARDENING_ENABLE="fortify stackprotector pic strictoverflow format relro bindnow"}
```
It seems that `=` instead of `:=` works, but is it documented?

cstrahan replied:
I could do that; do I then need to do something similar for bintools, though?

Ericson2314 replied:
Yeah if we go with two separate env vars, then it is not a problem.

orivej commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -52,6 +52,22 @@ rec {
+
+      hardeningEnable =
+        let val = attrs.hardeningEnable or [ ];
+        in if builtins.isList val then val else [ val ];
```
This is implemented in `lib.toList`.

cstrahan replied:
Nice find, thanks! I'll make that change shortly-ish.

orivej commented:
Why do you move hardening args before command line args? The latter will mess with hardening just by setting `-O1` or `-O0`.

globin commented:
Please don't merge yet, I want to go through this thoroughly and test if this still applies the hardening correctly (maybe add a checksecPhase using `checksec`)

cstrahan commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -3,6 +3,9 @@ set -o pipefail
 : ${outputs:=out}
 
+# If unset, assume the default hardening flags.
+: ${NIX_HARDENING_ENABLE="fortify stackprotector pic strictoverflow format relro bindnow"}
```
Yeah. Per the docs:

> When not performing substring expansion, using the form described below (e.g., ‘:-’), Bash tests for a parameter that is unset or null. Omitting the colon results in a test only for a parameter that is unset. Put another way, if the colon is included, the operator tests for both parameter’s existence and that its value is not null; if the colon is omitted, the operator tests only for existence.



https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html#Shell-Parameter-Expansion

orivej replied:
Interesting, this means we can write e.g. `${NIX_STORE-}` instead of `${NIX_STORE:-}`.

Ericson2314 replied:
Perhaps this should go in the setup-hook for cc-wrapper? `setup.sh` isn't exactly a paragon of clean layering, but I wouldn't want to make it worse.

cstrahan commented:
> Why do you move hardening args before command line args, now that hardening is not enabled outside nix-build environment? AFAIK programs often overspecify -O flags, and -O1 or -O0 will mess with hardening.

Assuming I interpreted the statements in #18995 correctly, some people were complaining about `-O` flags not working within builds, so I figured I'd try to address that.

As you point out, though, that does mean that the hardening would silently not apply if e.g. a Makefile sets `-O`. So, I'm thinking I should back out of that change, and within package definitions, people should just learn to unset the appropriate hardening flags if the target software requires, say, `-O` instead of `-O2` to work (and, of course, we should document that as a potential gotcha in the user guide).

I think I'll make that change soonish (hopefully several people can approve of and validate that idea).

And thanks for reviewing this :)

cstrahan commented:
I tried pretty hard to keep the behavior of `hardening{Enable,Disable}` the same as before. For example,

```
  hardeningEnable = [ "pie" ];
  hardeningDisable = [ "fortify" ];
```

should behave as before: add `"pie"` (which isn't/wasn't set by default), and remove `"fortify"` (which is/was otherwise set by default), while including all of the other default flags as before.

If there are any deviations that I'm missing, please do point them out.

cstrahan committed:
hardening: use lib.toList

cstrahan committed:
hardening: fix bug/typo

cstrahan commented:
/cc @Ralith

cstrahan committed:
hardening: fix careless bugs

I got a substitution backwards (used '+' instead of '-').

Also, this now works under `set -u` (had to fix a couple unbound
variable references).

dezgeg commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -107,6 +123,8 @@ rec {
             ++ optional (elem "host"   configurePlatforms) "--host=${stdenv.hostPlatform.config}"
             ++ optional (elem "target" configurePlatforms) "--target=${stdenv.targetPlatform.config}";
 
+        } // lib.optionalAttrs (hardeningDisable != [] || hardeningEnable != []) {
```
IIRC doing this sort of stuff on the Nix level is somewhat discouraged since it slows down evaluation of Nixpkgs.

Ericson2314 replied:
Well, more bash causes other problems too. I don't think its so clear cut.

Ralith commented:
> now that hardening is not enabled outside nix-build environment?

Does this change actually do that? I thought everyone was pretty much in agreement that `nix-shell` should always provide exactly the build environment, lest testing become a nightmare.

> As you point out, though, that does mean that the hardening would silently not apply if e.g. a Makefile sets -O. So, I'm thinking I should back out of that change, and within package definitions, people should just learn to unset the appropriate hardening flags if the target software requires, say, -O instead of -O2 to work (and, of course, we should document that as a potential gotcha in the user guide).

Almost every single person who wants to do C/C++ development with Nix will (continue to) be badly bitten if you leave things the way they currently are in master and have been historically, user guide or no. Completely and silently breaking debug info unless you set an esoteric configuration flag that nobody outside of Nix has any context for is an incredibly bad user experience. If packages are broken, fix the packages. I'm not even convinced there's a significant proportion of packages that have -O0 or -O1 hardcoded in their release builds in the first place without a good reason (or at all, for that matter), and when there is a good reason, silently clobbering it represents a frustrating landmine for packagers.

To be very clear, this goes for both `nix-shell` environments and `nix-build` itself. I often need to build dependencies with debug info, and Nix should not sabotage that.

orivej commented:
> Does this change actually do that? I thought everyone was pretty much in agreement that nix-shell should always provide exactly the build environment, lest testing become a nightmare.

This change lets standalone `gcc` not add hardening flags, in which case it no longer matters to move them before command line arguments. `nix-build` and `nix-shell` continue to make `gcc` add them by default.

Could you clarify how you build packages and their dependencies with debug info? Do you add `separateDebugInfo = true;` or `NIX_CFLAGS_COMPILE` to their definitions, or pass extra flags to `configure`/`make`/`cmake`, or modify their sources to build with debug info by default?

Ralith commented:
> This change lets standalone gcc not add hardening flags, in which case it no longer matters to move them before command line arguments

I'm not sure who's making non-trivial use of gcc outside of a `nix-shell` environment (or an actual package build), considering that having dependencies is useful, and indeed that letting you manage multiple distinct build environments with `nix-shell` is one of Nix's largest selling points to the developer audience.

> Could you clarify how you build packages and their dependencies with debug info? Do you add separateDebugInfo = true; or NIX_CFLAGS_COMPILE to their definitions, or pass extra flags to configure/make/cmake, or modify their sources to build with debug info by default?

When I'm working in `nix-shell`, i.e. compiling my project by invoking my build system by hand, I expect `gcc -O0 -g` to do what it does in every other environment in the world. When I want a debug build of a dependency, that usually means changing the dependency's build configuration (i.e. passing flags to `configure` or `cmake`) because there's often more to getting a "debug build" of a project than simply passing `-g`, such as (but not limited to) different sets of preprocessor definitions.

cstrahan commented:
@Ralith This continues to provide hardening support for both nix-build nix-shell. Outside of either of those two, NIX_HARDENING_ENABLE env var won't be set, and thus system-wide gcc will no longer default to enabling hardening.

Regarding optimization and debug builds: I only see three options:

1. Allow hardening to silently fail because the user and/or build system passed -O
2. Silently override the user/build-system supplied -O with -O2 when hardening is enabled
3. When the user supplied flag conflicts with the hardening flag, have the wrapper print an informative error and exit non-zero

At the present moment, this implements number 1, though as noted earlier I'm interested in moving to either of the latter two options.

Ralith commented:
> Outside of either of those two, NIX_HARDENING_ENABLE env var won't be set, and thus system-wide gcc will no longer default to enabling hardening.

Right, but a system-wide gcc isn't very useful to anyone regardless.

> At the present moment, this implements number 1, though as noted earlier I'm interested in moving to either of the latter two options.

Option 2 is what Nix has without this PR, and is why there is a long string of confused and unhappy developers in https://github.com/NixOS/nixpkgs/issues/18995, plus however many more who never managed to find the issue. It makes Nix hostile to developers who aren't already aware of the problem, and cost me personally a considerable amount of time, very nearly driving me to abandon Nix before I determined the cause.

3 would be tolerable; it's weird and nonstandard but you can't miss it, so it won't waste peoples' time the way 2 does, assuming the message is clear enough. It'd also make packages with broken builds obvious, which would clarify how common a problem that is.

bjornfor commented:
> Right, but a system-wide gcc isn't very useful to anyone regardless.

I think the point is that plain GCC provided by Nix should behave like any other GCC on any other distro. Injecting hardening flags is something that should be provided by the build environment, as environment variables or explicit command line options to the compiler.

Ralith commented:
I don't think using gcc outside of a Nix build environment is a very important case, frankly. I'm happy to see it no longer be affected by the issue, but that doesn't make the issue any less serious for the case I actually use.
A large part of the appeal of Nix for development is its ability to construct pure, immutable, reproducible, independent per-project build environments for you to work in; relying on external, globally installed tools is contrary to that paradigm. If installing things globally and switching them out every time I needed something different worked for me I'd still be on Arch.

cstrahan commented:
@fpletz @globin Ping.

As far as I'm concerned, the only thing this is waiting on is a decision on how to handle user supplied `-O` et al; see my comment here: https://github.com/NixOS/nixpkgs/pull/28029#issuecomment-321176884

cstrahan commented:
I don't know yet how I feel about this, but this popped into my head I figured I'd share it before I forget: I wonder if it might make sense to print the enabled hardening flags when first starting a `nix-shell`, and give a hint that they might want to tweak them with `hardening{Disable,Enable}`. The downside is that I personally don't like the thought of `nix-shell` spewing a bunch of text on each use, though that could be possibly addressed by having tools like `nix-shell` be verbose by default with some config mechanism to opt out. That's all very rough reasoning, and not directly related to this PR, but I figured I'd get that thought out in writing before I forget, FWIW.

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -1,32 +1,42 @@
 
-# Intentionally word-split in case 'hardeningDisable' is defined in Nix. The
+# Create table of unsupported flags for this toolchain.
+for flag in @hardening_unsupported_flags@; do
```
Do we still need the disabled map if `mkDerivation` converts everything to a whitelist? I rather have this filter the whitelist.

cstrahan replied:
Sounds good, will do :+1:

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -1,32 +1,42 @@
-for flag in ${hardeningDisable[@]} @hardening_unsupported_flags@
-do
-  hardeningDisableMap[$flag]=1
+for flag in ${NIX_HARDENING_ENABLE-}; do
```
Could we do `NIX_CC_WRAPPER_@infixSalt@_HARDENING_ENABLE`? For the sake of clean laying I've wanted to use the `NIX_CC_WRAPPER` prefix instead, so might as well start with a new environment variable. `add-flags.sh` can take care of dealing with the `@infixSalt@` part; it should be fairly easy to cargo cult.

cstrahan replied:
Will do.

cstrahan replied:
@Ericson2314 This still needs to be changed to `NIX_@infixSalt@_HARDENING_ENABLE` (under the assumption that we'll stick to the old name for now, rather than `NIX_CC_WRAPPER_HARDENING_ENABLE`), right?

I must confess that I haven't found the time to learn how your cross compiling changes work, though I hope to soon-ish. Of course "soon" for me probably translates to "in a couple eons" for most people, given my schedule as of late...

edolstra commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -1,32 +1,42 @@
-for flag in ${hardeningDisable[@]} @hardening_unsupported_flags@
-do
-  hardeningDisableMap[$flag]=1
+for flag in ${NIX_HARDENING_ENABLE-}; do
```
What is `infixSalt`? I've never seen it before, but all of a sudden Nixpkgs has dozens of references to it...

Ericson2314 replied:
@edolstra it's so we can have multiple cc-wrappers that don't conflict for cross compilation. I left a bunch of comments in https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/setup-hook.sh and https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/add-flags.sh

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -1,32 +1,42 @@
 
-# Intentionally word-split in case 'hardeningDisable' is defined in Nix. The
+# Create table of unsupported flags for this toolchain.
+for flag in @hardening_unsupported_flags@; do
```
Hmm, upon further reading, maybe something like:
```bash
if [[ -n "$NIX_DEBUG" ]]; then
    declare -A hardeningDisableMap=()
fi
for flag in @hardening_unsupported_flags@; do
    [[ -n ${hardeningEnableMap[$flag]} ]] | continue
    if [[ -n "$NIX_DEBUG" ]]; then
        hardeningEnableMap[$flag]=1
    fi
    unset hardeningEnableMap[$flag]
done
if [[ -n "$NIX_DEBUG" ]]; then
    // print disabled flags as before
fi
```
Then it is clear that all the logic follows from the whitelist, and the disable map is only used for debugging.

ElvishJerricco commented:
As far as https://github.com/NixOS/nixpkgs/pull/28029#issuecomment-321176884 goes, I'm also in favor of either solution 1 or 3. I'd say that if the breakage in `nixpkgs` with 3 is minimal, that's better. But it's not worth the effort of going through and fixing hundreds of derivations, so if it comes to that, 1 is probably better.

cstrahan commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -107,6 +119,8 @@ rec {
             ++ optional (elem "target" configurePlatforms) "--target=${stdenv.targetPlatform.config}";
 
+        } // lib.optionalAttrs (hardeningDisable != [] || hardeningEnable != []) {
+          NIX_HARDENING_ENABLE = enabledHardeningOptions;
```
@Ericson2314 Regarding your comment [regarding `NIX_CC_WRAPPER_@infixSalt@_HARDENING_ENABLE`](https://github.com/NixOS/nixpkgs/pull/28029/files#r133272454), what's the appropriate thing to do here? Should I mostly leave this alone (the only change, perhaps, being a rename to `NIX_CC_WRAPPER_HARDENING_ENABLE`), and then have (per another one of your recommendations) the setupHook for the CC-wrapper use the `$NIX_CC_WRAPPER_HARDENING_ENABLE` when `NIX_CC_WRAPPER_@infixSalt@_HARDENING_ENABLE` isn't defined? Curious what the exact logic should be here.

Ericson2314 replied:
1. https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/add-flags.sh#L7-L13 in there goes `NIX_CC_WRAPPER+HARDENING_ENABLE`.
2. this line becomes `NIX_CC_WRAPPER_HARDENING_ENABLE ..`

Now that there's a bintools wrapper too, we would have to do an analogous thing there with this plan. So OTOH if we want just one variable with both, then maybe we shouldn't add the `_CC_WRAPPER` to the name. *But*, no matter what we still want to do the infix salt.

Ralith commented:
It'd be really great if we could have this merged for 18.03 and be done with the issue.

oxij commented:
Agreed.

If @cstrahan is out of gas on this issue I can pick it up (I started implementing a similar thing independently, but this is better).

Also, I think nixpkgs gcc should fail when flags get overridden. Sure, the stuff will break, but its better to have stuff loudly broken than to have lingering security issues because we optimized something we shouldn't have.

dtzWill commented:
Friendly ping, let's not let this slip away unintentionally :).

fpletz commented:
I would also love to see this on 18.03. :+1: It would be great if someone could address the feedback and resolve the conflicts.

Ericson2314 commented:
I suppose I could fix the conflicts / resolve my review.

cstrahan commented:
I've rebased onto a recent commit to resolve conflicts.

cstrahan committed:
hardening: clarify the whitelist logic

Per @Ericson2314's suggestion [1], make it more clear that the active
hardenings are decided via whitelist; the blacklist is merely for the
debug messages.

1: https://github.com/NixOS/nixpkgs/pull/28029/commits/36d5ce41d4538e83199a000e6f849442c1cf959c#r133279731

cstrahan commented:
@Ericson2314 I addressed your code clarity suggestion in the latest commit (737390b) - how does it look now?

Also, did I get the infix salt related changes right?

cstrahan committed:
hardening: allow user supplied flags to override

Put hardening flags before user supplied flags.

cstrahan commented:
When I rebased, I undid the change to the ordering of hardening flags and user supplied flags. The last commit (7eea5bb7773cf28f892d79a88aa04571b5bb1696) changes things **such that user supplied flags override those from our hardening**. FWIW, I don't personally have strong feelings about this detail.

I think this is ready for (re)review at this point.

cstrahan commented:
I tried building `hello` to sanity check things, but that failed:

```
$ nix-build . -A hello
these derivations will be built:
  /nix/store/8a0cz8qzfg4v4l2yxmlk7icb4dvgxksr-stdenv.drv
  /nix/store/qlw07qppfjipzbwmfjdg67ymkpsyx0mk-hello-2.10.drv
building '/nix/store/8a0cz8qzfg4v4l2yxmlk7icb4dvgxksr-stdenv.drv'...
output '/nix/store/mhvcjf3rck7r907vn20dhis78kccngk5-stdenv' is not allowed to refer to the following paths:
    /nix/store/d1prcspbh2qsviipvnaxizcj8l3g7fpw-bootstrap-tools
    /nix/store/lan2w3ab1mvpxj3ppiw2sizh8i7rpz7s-busybox
cannot build derivation '/nix/store/qlw07qppfjipzbwmfjdg67ymkpsyx0mk-hello-2.10.drv': 1 dependencies couldn't be built
error: build of '/nix/store/qlw07qppfjipzbwmfjdg67ymkpsyx0mk-hello-2.10.drv' failed
```

I don't know why this is happening, though; I'd appreciate a helping hand.

cstrahan commented:
Rebasing onto master fixed the prohibited reference issue, so I think this is now working 100% again.

@Ericson2314 would you mind giving this final review, please?

cstrahan commented:
@Ericson2314 I need to do something similar in `pkgs/build-support/bintools-wrapper/add-hardening.sh`, right? And it looks like the types of hardening aren't the same between the cc-wrapper and bintools-wrapper, so maybe it would make sense to do as you suggested and have e.g. `NIX_CC_WRAPPER_HARDENING_ENABLE` and `NIX_BINTOOLS_WRAPPER_HARDENING_ENABLE`, instead of just `NIX_HARDENING_ENABLE`.

Ericson2314 commented:
@cstrahan great job with the disabled map! It's just the change I wanted.

Yes `bintools-wrapper` will need the same sort of change. I'm fine with two separate env vars or just `NIX_HARDENING_ENABLE`. See https://github.com/NixOS/nixpkgs/pull/28029#discussion_r164167528 for details re the infix salt suff.

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-hardening.sh:
```diff
@@ -1,33 +1,45 @@
+# Intentionally word-split in case 'NIX_HARDENING_ENABLE' is defined in Nix. The
+# array expansion also prevents undefined variables from causing trouble with
+# `set -u`.
+for flag in ${NIX_HARDENING_ENABLE-}; do
```
`s/NIX_/NIX_@infixSalt@_/`

Ericson2314 replied:
And the the corresponding `add-flags.sh` change I mentioned before.

cstrahan replied:
> `s/NIX_/NIX_@infixSalt@_/`

Working on that now.

> And the the corresponding add-flags.sh change I mentioned before.

Oh, I modified `pkgs/build-support/bintools-wrapper/add-flags.sh`. Do I want to modify both add-flags? I guess, with the split vars approach, I'd want to put the respective var name in each.

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -115,6 +115,19 @@ rec {
+        "fortify" "stackprotector" "pic" "strictoverflow" "format" "relro" "bindnow"
+      ];
+
+      hardeningDisable = lib.toList (attrs.hardeningDisable or [ ]);
```
These are already bound above :)

cstrahan replied:
The reason for this line (and the similar line below) is because I wanted to support the (at least historical) use case of passing a string (e.g. `"all"`) for these arguments, but to simplify logic down below, I wanted to coerce that value to a list.

At some point after I initially submitted this PR, it looks like we made the interface for `mkDerivation` such that `hardeningEnable` & `hardeningDisable` strictly *require* a list, rather than accepting a list *or* string. In light of that, yeah, these lines needn't exist :).

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -115,6 +115,19 @@ rec {
+
+      hardeningDisable = lib.toList (attrs.hardeningDisable or [ ]);
+
+      hardeningEnable = lib.toList (attrs.hardeningEnable or [ ]);
```
These are already bound above :)

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -115,6 +115,19 @@ rec {
         ]
       ];
 
+      defaultHardeningFlags = [
```
See also `supportedHardeningFlags` above. Just so you are aware.

cstrahan replied:
Ah, I could just subtract `{ "pie" }` from that set, if that's what you're hinting at.

cstrahan committed:
hardening: initial cross support

Ericson2314 commented on pkgs/build-support/bintools-wrapper/add-flags.sh:
```diff
@@ -24,10 +24,10 @@ if [ "${NIX_BINTOOLS_WRAPPER_@infixSalt@_TARGET_TARGET:-}" ]; then
 done
 for var in "${var_templates_bool[@]}"; do
-    mangleVarBool "$var" "${role_infixes[@]}"
+    mangleVarBool "$var" ${role_infixes[@]+"${role_infixes[@]}"}
```
Did you copy the `+` thing or did you find it was needed? No bash can tell the difference between empty and undefined arrays. Old bash needs your thing with `set -u` for empty and undefined alike, but new bash doesn't for empty and undefined alike.

cstrahan replied:
Yeah, got errors due to `role_infixes` being undefined. Happened while building `linuxHeaders`, though might have been poked due that in combination with some other element I'm missing. But yeah, this line failed otherwise.

cstrahan replied:
@dtzWill Yeah. Per the docs:

> When not performing substring expansion, using the form described below (e.g., ‘:-’), Bash tests for a parameter that is unset or null. Omitting the colon results in a test only for a parameter that is unset. Put another way, if the colon is included, the operator tests for both parameter’s existence and that its value is not null; if the colon is omitted, the operator tests only for existence.

https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html#Shell-Parameter-Expansion

dtzWill replied:
TIL! Thanks! :+1:

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-flags.sh:
```diff
@@ -10,6 +10,7 @@ var_templates_list=(
     NIX+CXXSTDLIB_COMPILE
     NIX+CXXSTDLIB_LINK
     NIX+GNATFLAGS_COMPILE
+    NIX+HARDENING_ENABLE
```
Per https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L20-L23 you should put a variable used by both in either just `bintools-wrapper` or both wrappers, never just `cc-wrapper`.

cstrahan replied:
I've removed this line in the latest commit.

cstrahan committed:
hardening: ld wrapper changes, setup-hook, etc

cstrahan commented on pkgs/build-support/cc-wrapper/setup-hook.sh:
```diff
@@ -147,6 +147,10 @@ export ${role_pre}CXX=@named_cxx@
 export CXX${role_post}=@named_cxx@
 
+# If unset, assume the default hardening flags.
+: ${NIX_HARDENING_ENABLE="fortify stackprotector pic strictoverflow format relro bindnow"}
```
When I split up into two variables, this will contain just the flags that make sense for the cc wrapper. Similarly for the bintools' setup-hook.

cstrahan replied:
(For posterity: per IRC convo, I not sure if we're going with the split variables over the unified variable approach. In light of this, the above comment might be misleading.)

cstrahan commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -74,6 +74,11 @@ rec {
+      enabledHardeningOptions =
+        if builtins.elem "all" hardeningDisable
+        then []
+        else lib.subtractLists hardeningDisable (defaultHardeningFlags ++ hardeningEnable);
```
Hopefully a little cleaner than before.

cstrahan committed:
hardening: make requested fixes

cstrahan commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -7,7 +7,6 @@ fi
 
 : ${outputs:=out}
 
-
```
No, that was an accident. I've reverted this change in the latest commit.

cstrahan commented:
@Ericson2314 I think I've incorporated all of your feedback. Could I ask you to review this, please? Getting this in (hopefully sooner than later) would make my life a lot easier.

Ericson2314 commented:
@cstrahan looking now

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -74,6 +74,11 @@ rec {
+        then []
+        else lib.subtractLists hardeningDisable (defaultHardeningFlags ++ hardeningEnable);
       # hardeningDisable additionally supports "all".
       erroneousHardeningFlags = lib.subtractLists supportedHardeningFlags (hardeningEnable ++ lib.remove "all" hardeningDisable);
```
@cstrahan does this overlap with the `lib.subtractLists` above?

cstrahan replied:
@Ericson2314  Overlap in what way? I think they're distinct uses (one to check for erroneous flags, the other to calculate the effective flags).

Ericson2314 commented:
@cstrahan that's my only question. Otherwise looks good!

Ericson2314 reviewed (approved):
I'll worry about this nix-level stuff after.

Ericson2314 commented on pkgs/build-support/bintools-wrapper/add-hardening.sh:
```diff
@@ -1,33 +1,45 @@
+for flag in @hardening_unsupported_flags@; do
+  [[ -n ${hardeningEnableMap[$flag]} ]] || continue
+  if (( "${NIX_DEBUG:-0}" >= 1 )); then
+    hardeningDisableMap[$flag]=1
```
This only matters for `hardening_unsupported_flags` that are not in `allHardeningFlags`. Do we care?

cstrahan replied:
You mean "This only matters for hardening_unsupported_flags that *are* in allHardeningFlags", right? The idea is that if the toolchain doesn't support a given flag, we remove it from the set of enabled flags -- are you questioning the utility of that?

cstrahan replied:
Oh, you're drawing a comparison to the loop over `allHardeningFlags` below.

Ericson2314 replied:
Yes, the latter. Sorry that was unclear.

cstrahan replied:
I suppose we don't really need to report unsupported flags as if they were disabled. Should I remove this conditional (and its body)?

cstrahan committed:
hardening: simplify reporting of disabled flags

Ericson2314 committed:
hardenning: Rejigger ifs and explicit declare and unset -v

Ericson2314 committed:
hardening: Reindent

Ericson2314 committed:
hardening: line order, spacing, and pointless quoting for consistency

Ericson2314 reviewed (changes requested):
As I talked to @cstrahan on IRC, I made https://github.com/cstrahan/nixpkgs/pull/2 with some things. Of course we can leave out the more cosmetic stuff like reindenting. But I think the changes in control flow are meaningful.

Ericson2314 committed:
hardening: Tiny reindent

Ericson2314 commented:
Damnit. On the phone it hides the base branch.

Ericson2314 commented:
If someone beats me to it, please send to staging instead.

matthewbauer commented:
Reverted in master. Haven't done anything in staging.

Ericson2314 commented:
@matthewbauer thanks so much! Reverted the revert in on staging so we're all good now.

Ericson2314 commented:
@cstrahan thanks again for making this, and sorry it took so long to get it merged.

oxij commented:
Fun fact: I did #39463 and other such fixes before this PR got merged and now I rebuilt everything with #39464 again. And several new packages fail their tests because of hardening. A couple of examples (apply #39464 and try building those to confirm):

```diff
diff --git a/pkgs/development/libraries/a52dec/default.nix b/pkgs/development/libraries/a52dec/default.nix
index d8a56a3d28e..264e8795f52 100644
--- a/pkgs/development/libraries/a52dec/default.nix
+++ b/pkgs/development/libraries/a52dec/default.nix
@@ -1,24 +1,28 @@
 {stdenv, fetchurl}:
 
 stdenv.mkDerivation rec {
   name = "a52dec-0.7.4p4";
-  
+
   src = fetchurl {
     url = "${meta.homepage}/files/a52dec-0.7.4.tar.gz";
     sha256 = "0czccp4fcpf2ykp16xcrzdfmnircz1ynhls334q374xknd5747d2";
   };
 
   # From Handbrake
   patches = [
     ./A00-a52-state-t-public.patch
     ./A01-thread-safe.patch
     ./A02-imdct-shutup.patch
     ./A03-automake.patch
   ];
 
+  hardeningDisable = [ "format" ];
+  # # with the above removed
+  # doCheck = false; # fails 1 out of 1 tests with "BAD GLOBAL SYMBOLS"
+
   meta = {
     description = "ATSC A/52 stream decoder";
     homepage = http://liba52.sourceforge.net/;
     platforms = stdenv.lib.platforms.unix;
   };
 }
diff --git a/pkgs/development/libraries/ilmbase/default.nix b/pkgs/development/libraries/ilmbase/default.nix
index 956a8667be5..b088414d377 100644
--- a/pkgs/development/libraries/ilmbase/default.nix
+++ b/pkgs/development/libraries/ilmbase/default.nix
@@ -9,20 +9,26 @@ stdenv.mkDerivation rec {
   };
 
   outputs = [ "out" "dev" ];
 
   preConfigure = ''
     ./bootstrap
   '';
 
   buildInputs = [ automake autoconf libtool which ];
 
+  hardeningDisable = [ "format" ];
+  # # with the above removed
+  # # fails 1 out of 1 tests with
+  # # "lt-ImathTest: testBoxAlgo.cpp:892: void {anonymous}::boxMatrixTransform(): Assertion `b21 == b2' failed"
+  # doCheck = false;
+
   NIX_CFLAGS_LINK = [ "-pthread" ];
 
   patches = [ ./bootstrap.patch ];
 
   meta = with stdenv.lib; {
     homepage = http://www.openexr.com/;
     license = licenses.bsd3;
     platforms = platforms.all;
     maintainers = with maintainers; [ wkennington ];
   };
```

oxij commented:
Okay, after some more investigation this is not as scary as I though. I mixed up i686 and x86_64 evaluations a bit, so

> pkgs/development/libraries/a52dec/default.nix

only needs

```
  hardeningDisable = stdenv.lib.optional stdenv.isi686 "pic";
  # with the above removed
  # doCheck fails 1 out of 1 tests with "BAD GLOBAL SYMBOLS"
```

to pass (and that's likely a bug in the tests themselves) and ilmbase seems to be broken on i686 (which is kinda scary since wine depends on it), but it's not hardenings' fault.

It is still the case that many packages broke with `-Wformat-security`, but I got sweating here thinking that format hardening could make a difference in tests. But it does not. Good.

Then, I think, this PR needs no follow-ups in that regard.

Ralith commented:
```
cc1: error: -Wformat-security ignored without -Wformat [-Werror=format-security]
cc1: some warnings being treated as errors
```

Bad combination of flags, somehow?

oxij commented:
GHC is not the only package that has this issue. I have a whole bunch of fixes for exactly that error message. Will PR in the next couple of days.

Ericson2314 commented on pkgs/build-support/bintools-wrapper/ld-wrapper.sh:
```diff
@@ -57,8 +57,8 @@ fi
 
-extraAfter=("${hardeningLDFlags[@]}")
-extraBefore=()
+extraAfter=()
```
@oxij @FRidh if we do all of this PR *but* this part, we should have the same behavior as before. That seems like a good start to me

FRidh replied:
Sounds good.

Ericson2314 replied:
I really need to go to sleep but somebody else feel free to try it.

Ralith replied:
Can the hardening flags be injected at the start instead of the end? I'd hate to see us reintroduce the problems that prompted this PR in the first place.

ElvishJerricco commented:
Can we please revert this on master until the *many* broken packages are fixed on staging?

oxij commented:
I fixed everything I know of on master #41192

cstrahan commented:
The breakage is mostly, if not entirely, due to the change in the order in which the flags are inserted. If the breakage is too great to keep things as they are now, I would recommend committing the couple line diff needed to revert just that part of this change set. That would retain 99% of the benefits while fixing any currently broken packages.
